clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3803,11 +3803,11 @@ void handleResponse(AbstractResponse abstractResponse) {
                     }
 
                     if (!partitionsWithErrors.isEmpty()) {
-                        partitionsToQuery.keySet().retainAll(partitionsWithErrors);
                         Set<String> retryTopics = partitionsWithErrors.stream().map(tp -> tp.topic()).collect(Collectors.toSet());
+                        topicPartitionOffsets.keySet().retainAll(partitionsWithErrors);
```

Comment:
Hmm.. I don't think it's safe to modify `topicPartitionOffsets`. This is shared by multiple calls following the initial `getListOffsetsCalls`. It would be better to create a new map. It might bee worth having a test case which uses two partitions with different leaders to verify this case is handled correctly.

Reply:
That's a good point, there's a possibility of clobbering of data between calls. I just pushed a change for this as well as added a test as you suggested that starts out with partitions with the same leader, refreshes metadata, and moves to different leaders.